### ISSUE TITLE
partial lambda breaks subsequent macro

### ISSUE THREAD

**Issue by matter123** [jeff-hykin/cpp-textmate-grammar#136]
example code
```c++
#define TEST(name) _ts->test(name) = [&]() -> bool

#define ASSERT(expected, actual)                                                        \
```
image:
![Screenshot from 2019-04-27 23-07-13](https://user-images.githubusercontent.com/714007/56859533-400e9900-6941-11e9-8912-8b03d7ea3cb6.png)

The `#define` does have the scope `meta.preprocessor.macro.cpp`

**Comment by jeff-hykin** [comment]
Things that need fixing
- Lambdas
- Function definition
- class
- struct
- enum
- union
- namespace
- (in the future) control flow

**Comment by jeff-hykin** [comment]
Conceptually, I'm splitting the preprocessor issues into two. 1. is the conditional logic (which I think we have a good plan for) and the other being this issue (incomplete pattern-ranges in macros).

To fix this one, I think we're going to have to make separate "macro-safe" versions of everything. The macro context will only include the macro safe versions, and all of the pattern-ranges in the macro version will have to end with a lookahead for a newline that doesn't have `\`. It won't be easy, but once it's done I think it'll solve all of the problems and be tagged correctly.

**Comment by matter123** [comment]
Either the macro safe versions have to be really careful not to overshoot, or the `#define` pattern can break up the macro line-by-line but that breaks patternRanges.

**Comment by jeff-hykin** [comment]
Yeah, I'm thinking maybe we can do it functionally, where every single pattern in the repo is copied and all of them that are ranges get a lookahead automatically applied to them

**Comment by jeff-hykin** [comment]
I created an initial fix for this which was surprisingly easy to write and will probably cover 80% of the multi-line cases. Between the single-line macro pattern and this shallow multi-line fix, 99% of the use cases should be covered. A full fix should still be implemented eventually, but its probably fine to wait until everything else has been fixed.

A full fix for this (recursively adding an ending pattern to all included patterns) will also fix the major issue for embedding other languages. The entire other language can be imported and every pattern range in the language can have a lookahead recursively appended to it.

**Comment by matter123** [comment]
What kind of cases would not be covered by your fix?

**Comment by jeff-hykin** [comment]
yeah, and that includes anything that is not directly in the $initial_context, e.g. everything in the evaluation context

**Comment by matter123** [comment]
So the 100% case would be to duplicate every pattern. Then change every include in those patterns to refer to the duplicated ones?

Or for convenience, duplicate the whole grammar, add the early bailout. And #define embeds the grammar.

**Comment by jeff-hykin** [comment]
That sounds like an easier solution. We can use the `recursively_set_each_value!` function to find all of the end patterns in the entire grammar and add the early bailout. This would work quite well for language embeddings too

**Comment by jeff-hykin** [comment]
I removed the macro_argument from root but theres a minor regression with the :vararg_ellipses

**Comment by matter123** [comment]
If there is going to be support for early bailout when embedded, then they should probably be in a separate scope, maybe `source.cpp.embedded.macro` and `source.cpp.embedded.latex`.

My concern is that `cpp.tmLanguage.json` is quite large, about 15k line or 518kb currently. Adding support for early bailout inside the existing grammar would nearly triple those numbers.

**Comment by jeff-hykin** [comment]
For embedding I was actually thinking it would be the main embed-er language that would have to deal with all those extra lines. Instead of doing "source.cpp" they'd have something like `grammar[:cpp_lang] = IO.read("cpp_grammar.json")` And then after importing it they'd add their own custom bailout pattern. It would add ~19K lines to their tmLanguage, but it would fix the behavior. For a language like markdown, this would still explode the grammar size.

Honestly, I'm surprised the size of grammar isn't causing more performance issues already.

**Comment by matter123** [comment]
c++ is by far the largest default grammar, it is over 3 times larger than the second largest grammar.

If a modification is needed to it, so that it can be properly embedded, the modified c++ syntax would 
accounts for, on average, over 85% of the syntax.
<details>
<summary># of Lines in default grammars (42 rows)</summary>

Language | Lines
---------|------
bat | 738
clojure | 435
coffeescript| 1,316
c | 3,881
**cpp** | **19,416**
csharp | 4,333
docker | 102
fsharp | 1,751
go | 676
groovy | 1,383
handlebars | 852
hlsl | 213
html | 2,643
ini | 113
java | 1,622
javascript | 5,592
json | 213
latex | 2387
less | 542
log | 93
lua | 281
make | 643
markdown | 2,569
objective-c | 3,606
perl | 315
php | 3,335
powershell | 1007
pug | 987
python | 5,279
r | 548
ruby | 2,775
rust | 692
sass | 1,730
shaderlab | 204
shell | 1,283
sql | 519
swift | 3,008
typescript | 5592
vb | 238
xml | 387
yaml | 621

</details>

**Comment by jeff-hykin** [comment]
well its good to see that we're winning, kinda 😂

**Comment by jeff-hykin** [comment]
Wow the C++ `:function_definition` pattern is larger than the entire go syntax
